extension/prometheus/prometheus.go

```
@@ -0,0 +1,101 @@
+		notificationProcessingDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Name:      "notification_processing_duration_seconds",
+				Help:      "histogram of notifications handled latencies",
+				Buckets:   []float64{0.1, 0.5, 0.9, 0.99}, //buckets for histogram
+			},
+			[]string{"success", "retry"},
+		),
+	}
+}
+
+// RegisterMetrics returns http handler for prometheus
+func (m *Metrics) RegisterMetrics(registry *prometheus.Registry) error {
+	err := registry.Register(m.notificationCounter)
+	if err != nil {
+		return err
+	}
+
+	err = registry.Register(m.notificationQueueDuration)
+	if err != nil {
+		return err
+	}
+
+	return registry.Register(m.notificationProcessingDuration)
+}
+
+// ReceivedNotification counts received notifications
+func (m *Metrics) ReceivedNotification(isNotification bool) {
+	labels := prometheus.Labels{"is_notification": strconv.FormatBool(isNotification)}
+	m.notificationCounter.With(labels).Inc()
+}
+
+// QueueNotification returns http handler for prometheus
+func (m *Metrics) QueueNotification(notification *sql.ProjectionNotification) {
+	key := "q" + fmt.Sprintf("%p", notification)
+	m.notificationStartTimes.Store(key, time.Now())
+}
+
+// StartNotificationProcessing is used to record start time of notification processing
+func (m *Metrics) StartNotificationProcessing(notification *sql.ProjectionNotification) {
+	key := "p" + fmt.Sprintf("%p", notification)
+	m.notificationStartTimes.Store(key, time.Now())
+}
+
+// FinishNotificationProcessing is used to observe end time of notification queue and processing time
+func (m *Metrics) FinishNotificationProcessing(notification *sql.ProjectionNotification, success bool, retry bool) {
+	memAddress := fmt.Sprintf("%p", notification)
+	queueStartTime, _ := m.notificationStartTimes.Load("q" + memAddress)
+	processingStartTime, _ := m.notificationStartTimes.Load("p" + memAddress)
```

Please do not ignore errors here. If the start time can't be loaded we should not send the related metrics.